submariner-operator/diagnose-refactor-ocm.md

```diff
@@ -0,0 +1,151 @@
+# Refactor subctl diagonose for use with OCM
+
+## Summary
+
+Currently only way to run diagnostics on a Submariner installation is through `subctl` CLI. This requires user or admin
+to run it manually from a host. There is no way to run it through UI like OCM or deployment tools like helm charts etc.
+
+## Proposal
+
+This enhancement proposes a means to run `subctl diagnose` through a `SubmarinerDiagnose` Job. User or Admin can deploy
+this Job manually. In case of OCM, a new SubmarinerDiagnose Controller will be added to SubmarinerAddon which will deploy
+this Job, track its completion and fetch results once done.
+
+## Design Details
+
+* A new option `--output json` will be added to `subctl diagnose` to capture output in a json format instead of current
+ user readable text format.
+* `subctl diagnose` will also update some Prometheus metrics whenever it is run.
+* A `submariner-diagnose` Job will be created which will run `subctl diagnose` with relevant flags as `command`. This
+ job will run once to completion. This Job will always run with new option of `--output json`.
+* A new `SubmarinerDiagnose` CRD will be added to SubmarinerAddon. OCM UI will create this CR with relevant values
+ to trigger running diagnose from OCM UI.`
+* A new `SubmarnerDiagnoseController` Controller will be added to OCM SubmarinerAddon. This controller will spawn the SubmarinerDiagnose
+ Job when SubmarinerDiagnose CR is created, track completion.
```

Comment:
With https://github.com/submariner-io/submariner-operator/pull/1790, diagnose can be invoked programmatically instead of the additional overhead of creating a `Job` with a new image. The output can be captured via a `reported.Interface` instance.

Reply:
UI code can't use Go APIs. So we need something to proxy UI requests. To use API we will need to add implementation in submariner-addon, which was the first iteration draft we had. UI can easily show the pod logs directly as it can use any K8s restful API.

Reply:
ok so we don't actually need the `SubmarnerDiagnoseController` as this states. But if we were to use the `SubmarnerDiagnoseController` it could use the **submariner-operator** API instead of creating a `Job`.

Reply:
Yes, the idea here is to avoid tying the UI to `Job`; instead, the UI only knows how to ask the operator to start a diagnostic, and we can change the operator to use something other than a `Job` if appropriate later on without changing the UI.